src/Avalonia.Controls/Control.cs

```diff
@@ -327,15 +333,21 @@ internal void OnLoadedCore()
         /// </summary>
         internal void OnUnloadedCore()
         {
-            if (_isLoaded)
+            switch (_loadState)
```

Comment:
Note: code could have been changed to simply what is below. That would be a much smaller change and perhaps simplify in some ways.

I do like thinking about this as 3-state logic with a LoadPending state though. It helps on other areas.

```csharp

// Remove from the loaded event queue here as a failsafe...
_loadedQueue.Remove(this)

if (_isLoaded)
{
    _isLoaded = false;
    OnUnloaded(...);
}

```

Reply:
I wanted to avoid the `HashSet` lookup in standard cases (going from `Loaded` to `Unloaded`, which should be the 99% path).

Reply:
That's true. It's impossible to over-optimize this too since it's for every control.